File: DSU.java
public class DSU{
    // Declaring two arrays to hold parent and rank of each node
    private int[] parent;
    private int[] rank;

    DSU(int n){
        /*
         *  Constructor
         *  */
        parent = new int[n];
        rank = new int[n];

        //initializing their values
        for (int i = 0; i < n; i++){
            parent[i] = i;
            rank[i] = 0;
        }
    }

    public int find(int node){
        /*
        *  Find function: searches for the parent of the particular node
        *  If node is the parent if itself it is the leader of the tree
        *  If not, uses path compression to find parent
        * */
        if(node == parent[node]) return node;

        return parent[node] = find(parent[node]);
    }

    public void Union(int u, int v){
        /*
        * Union Function: merge the set to which u belongs to and the set
        * to which v belongs.
        *
        * It attaches lower rank trees to high ranks trees while merging,
        * if ranks are equal it increases teh rank of the tree to which
        * other is joined by one
        *
        * */
        u = find(u);
        v = find(v);

        if (u != v){
            if (rank[u] < rank[v]){
                int temp = u;
                u = v;
                v = temp;
            }

            parent[v] = u;

            if (rank[u] == rank[v])
                rank[u]++;
        }
    }
}

File: Union-FInd_algorithms/DSU.py
class DSU:
    def __init__(self, n):
        """
            Declare two arrays to hold parent and node information
        """
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]

    def find(self, node):
        """
            Finds the parent of a node
            :param node:
            :return: Parent node
        """
        if node == self.parent[node]:
            return node
        self.parent[node] = self.find(self.parent[node])
        return self.parent[node]

    def union(self, u, v):
        u = self.find(u)
        v = self.find(v)

        if u != v:
            if self.rank[u] < self.rank[v]:
                temp = u
                u = v
                v = temp

                self.parent[v] = u
                if self.rank[u] == self.rank[v]:
                    self.rank[u] += 1
